### PR TITLE
Add UB rules

### PR DESCRIPTION
**Pull request by stoletheminerals**
Adding two UB-related rules. 
@cdesouza-chromium can you check if tests are correct?

### CODE REVIEW

**Comment by thypon** [comment]
The void* might trigger a bit too much

**Comment by stoletheminerals** [comment]
In brave-core we have "only" 47 uses of `void*`, this rule shouldn't trigger on `void` or `void *`, should be fine imo

**Comment by cdesouza-chromium** [comment]
@stoletheminerals Looking good!

**Review by cdesouza-chromium — APPROVED** [review]

**Comment by cdesouza-chromium** [comment]
> The void* might trigger a bit too much

To complement what has been said, `void` and `void*` are different. `void*` may be necessary in some rare cases, but it means a pointer without a type.

File: assets/semgrep_rules/client/reinterpret_cast.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        thypon
+        cdesouza-chromium
+    languages: [cpp]
+    message: "Using `reinterpret_cast` against some data types may lead to undefined bheaviour. In general, when needing to do these conversions, check how Chromium upstream does them. Most of the times a reinterpret_cast is wrong and there's no guarantee the compiler will generate the code that you thought it would."
+    severity: WARNING
+    patterns:
+      - pattern: reinterpret_cast<$T>($ARG)
+      - metavariable-regex:
+          metavariable: $T
+          regex: ^(.*int.*|.*double.*|.*float.*|.*char.*)$ # this probably needs to be tweaked
```
Comment (cdesouza-chromium) [inline assets/semgrep_rules/client/reinterpret_cast.yaml]: I think we should add `size_t` too to this list.

Comment (stoletheminerals) [reply assets/semgrep_rules/client/reinterpret_cast.yaml]: Added in https://github.com/brave/security-action/pull/690/commits/6b228f39d4eccd93c455233e0b16043a1f5d59e2

**Review by thypon — APPROVED** [review]
looking good, feel free to merge

**Pull request merged**